### PR TITLE
desktop shortcut support for windows

### PR DESCRIPTION
**Pull request by roenyroeny**
implemented the windows path for desktop shortcut creation of titles.

Icons are saved in the yuzu appdata folder under /icons

### CODE REVIEW

File: src/yuzu/main.cpp [hunk 1]
```diff
@@ -3789,6 +3805,36 @@ bool GMainWindow::CreateShortcut(const std::string& shortcut_path, const std::st
...
+    pShellLink->SetArguments(warguments.data());
+    pShellLink->SetDescription(wcomment.data());
+    pShellLink->SetIconLocation(wicon_path.data(), 0);
+
+    IPersistFile* pPersistFile;
+    hres = pShellLink->QueryInterface(IID_IPersistFile, (void**)&pPersistFile);
+    if (FAILED(hres))
+        return false;
+
+    hres = pPersistFile->Save(wshortcut_path.data(), TRUE);
```
Comment (FearlessTobi) [inline src/yuzu/main.cpp:3830]: dito

File: src/yuzu/util/util.cpp [hunk 2]
```diff
@@ -37,3 +38,81 @@ QPixmap CreateCirclePixmapFromColor(const QColor& color) {
...
+
+    iconFile.close();
+
+    return true;
+}
+#else
+bool SaveAsIco(QImage image) {
+    return false;
+}
+#endif
```
Comment (FearlessTobi) [inline src/yuzu/util/util.cpp:118]: needs a newline here

File: src/yuzu/util/util.cpp [hunk 3]
```diff
@@ -37,3 +38,81 @@ QPixmap CreateCirclePixmapFromColor(const QColor& color) {
     painter.drawEllipse({circle_pixmap.width() / 2.0, circle_pixmap.height() / 2.0}, 7.0, 7.0);
     return circle_pixmap;
 }
+
+#if defined(WIN32)
+#include <Windows.h>
```
Comment (FearlessTobi) [inline src/yuzu/util/util.cpp:43]: this include should be defined at the top of the file

**Review by FearlessTobi — CHANGES_REQUESTED** [review]

**Comment by FearlessTobi** [comment]
Please also fix clang-format.

File: src/yuzu/main.cpp [hunk 4]
```diff
@@ -2689,7 +2691,7 @@ void GMainWindow::OnGameListCreateShortcut(u64 program_id, const std::string& ga
             QMessageBox::critical(
                 this, tr("Create Shortcut"),
                 tr("Cannot create shortcut on desktop. Path \"%1\" does not exist.")
-                    .arg(QString::fromStdString(target_directory)),
```
[the remark this answers is not in the document]
Comment (roenyroeny) [reply src/yuzu/main.cpp:2692]: target_directory is a std::filesystem::path which cannot implicitly be converted to a std::string in MSVC

**Comment by DanielSvoboda** [comment]
Take a look at my version, I hadn't seen what you had started doing, it's fully functional

https://github.com/yuzu-emu/yuzu/compare/master...DanielSvoboda:yuzu:master

Some details:

remove_special_characters
Removes all characters that Windows does not accept: '\\', '/', ':', '?', '"', '<', '>', '|'
Not just the ':' as it currently is
std::replace(title.begin(), title.end(), ':', '-');

savePixmapsToICO
This code implements a custom function to convert PNG to ICO and save using QPixmap.
The code has a lot of additional details to handle various situations and ensure that images are saved correctly in the .ico file
Save the icons in the AppData\Roaming\yuzu\icons folder, an equivalent to the path used in Linux

Uses the Windows API function SHGetFolderPath to get the path of the current desktop directory.
Accept directories with accents like my case, where in my language the desktop is called "Área de trabalho"
It uses the COM (Component Object Model) interfaces `IShellLink` and `IPersistFile` to create a shortcut (file.lnk)

**Comment by boludoz** [comment]
> Please review my version. I had not seen what you had started working on, but it is fully functional.
> 
> Some details:
> 
> ```
> remove_special_characters removes all characters that Windows does not accept, not just the colon (:).
> This code implements a custom function to convert PNG to ICO and save it using QPixmap.
> The code has many additional details to handle various situations and ensure that images are saved correctly in the .ico file.
> ```
> 
> Here is the complete code: [master...DanielSvoboda:yuzu:shortcut](https://github.com/yuzu-emu/yuzu/compare/master...DanielSvoboda:yuzu:shortcut)

Give the guy a chance, he's a software engineer, his code seems good and neat, with good support for characters.

**Comment by boludoz** [comment]
> Please review my version. I had not seen what you had started working on, but it is fully functional.
> 
> Some details:
> 
> ```
> remove_special_characters removes all characters that Windows does not accept, not just the colon (:).
> This code implements a custom function to convert PNG to ICO and save it using QPixmap.
> The code has many additional details to handle various situations and ensure that images are saved correctly in the .ico file.
> ```
> 
> Here is the complete code: [master...DanielSvoboda:yuzu:shortcut](https://github.com/yuzu-emu/yuzu/compare/master...DanielSvoboda:yuzu:shortcut)

Give him a chance, he is an excellent coder and knows what he is doing since he is a software engineer with work experience.

**Comment by roenyroeny** [comment]
> Please review my version. I had not seen what you had started working on, but it is fully functional.
> 
> Some details:
> 
> ```
> remove_special_characters removes all characters that Windows does not accept, not just the colon (:).
> This code implements a custom function to convert PNG to ICO and save it using QPixmap.
> The code has many additional details to handle various situations and ensure that images are saved correctly in the .ico file.
> ```
> 
> Here is the complete code: [master...DanielSvoboda:yuzu:shortcut](https://github.com/yuzu-emu/yuzu/compare/master...DanielSvoboda:yuzu:shortcut)

Looks alright to me. Do you want to merge in your remove_special_characters and merge this PR as yours is closed ?
Excuse the delay btw, i haven't had much time to look at this lately

**Comment by DanielSvoboda** [comment]
you can use any part of my code

**Comment by boludoz** [comment]
I recommend giving full support for u8string

**Comment by FearlessTobi** [comment]
Looks much better now, however to forgot to address a few of my review comments.

**Comment by boludoz** [comment]
```
// Illgal and null characters remover for name: Windows, MacOS and Linux
// 1. Full u8string support
// 2. Remember: dont use prohibited chars in ReplaceString
void NameSantizer(std::u8string &u8filename, std::u8string ReplaceString = u8"_") {
  std::u8string u8path_santized;

  for (auto &c : u8filename) {
    if (c == u8'\\' || c == u8'/' || c == u8':' || c == u8'*' || c == u8'?' ||
        c == u8'\"' || c == u8'<' || c == u8'>' || c == u8'|' || c == u8'\0') {
      u8path_santized += ReplaceString;
      continue;
    }

    if (c == u8'.') {
      u8path_santized += ReplaceString;
      continue;
    }

    [[likely]]u8path_santized += c;
  }

  // Delete all spaces and dots at the end
  while (u8path_santized.back() == u8' ' || u8path_santized.back() == u8'.') {
    u8path_santized.pop_back();
  }

  if (u8path_santized.empty()) {
    throw std::runtime_error("[NameSantizer] Empty filename or all chars are wrong");
  }

  u8filename = u8path_santized;

}
```

**Comment by boludoz** [comment]
And [DanielSvoboda](https://github.com/DanielSvoboda) made a very good function with qt for ico files, that has better cross-platform support.

**Comment by boludoz** [comment]
`#include <filesystem>
#include <fstream>
#include <iostream>
#include <string>

// Illgal and null characters remover for name: Windows, MacOS and Linux
// Special case for ':' in games
// 1. Full u8string support
// 2. Remember: dont use prohibited chars in ReplaceString
void NameSantizer(std::u8string &u8filename) {
  std::u8string u8path_santized;

  u8path_santized = u8filename;

  int eSizeSanitized = u8path_santized.size();

  // Special case for ":", for example: 'Pepe: La secuela' --> 'Pepe - La
  // secuela' or 'Pepe : La secuela' --> 'Pepe - La secuela'
  for (size_t i = 0; i < eSizeSanitized; i++) {

    switch (u8path_santized[i]) {
    case u8':':
      if (i == 0 || i == eSizeSanitized - 1) {
        u8path_santized.replace(i, 1, u8"_");
      } else if (u8path_santized[i - 1] == u8' ') {
        u8path_santized.replace(i, 1, u8"-");
      } else {
        u8path_santized.replace(i, 1, u8" -");
        eSizeSanitized++;
      }
      break;
    case u8'\\':
      [[fallthrough]];
    case u8'/':
      [[fallthrough]];
    case u8'*':
      [[fallthrough]];
    case u8'?':
      [[fallthrough]];
    case u8'\"':
      [[fallthrough]];
    case u8'<':
      [[fallthrough]];
    case u8'>':
      [[fallthrough]];
    case u8'|':
      [[fallthrough]];
    case u8'\0':
      u8path_santized.replace(i, 1, u8"_");
      break;
    default:
      break;
    }
  }

  // Delete duplicated spaces || Delete duplicated dots (MacOS i think)
  for (size_t i = 0; i < eSizeSanitized; i++) {
    if ((u8path_santized[i] == u8' ' && u8path_santized[i + 1] == u8' ') ||
        (u8path_santized[i] == u8'.' && u8path_santized[i + 1] == u8'.')) {
      u8path_santized.erase(i, 1);
      i--;
    }
  }

  // Delete all spaces and dots at the end (Windows almost)
  while (u8path_santized.back() == u8' ' || u8path_santized.back() == u8'.') {
    u8path_santized.pop_back();
  }

  if (u8path_santized.empty()) {
    throw std::runtime_error(
        "[NameSantizer] Empty filename or all chars are wrong");
  }

  u8filename = u8path_santized;
}

int main() {
  std::u8string utf8String =
      u8"Pepe: la secuela"; // Tu cadena de texto en formato u8string
  NameSantizer(utf8String);
  std::filesystem::path filePath = utf8String;
  std::ofstream outputFile(filePath, std::ios::binary);

  if (!outputFile) {
    return 1;
  }
  std::string utf8StringAsStdString(utf8String.begin(), utf8String.end());
  outputFile.write(utf8StringAsStdString.c_str(), utf8StringAsStdString.size());
  outputFile.close();
  return 0;
}
`
Hello sir, I made a code to fix all special and null characters, it just doesn't take into account the case of special Windows names, but don't worry, that will never happen.
Additionally, correct the names of the games so that they look prettier if they contain ':', if you give me permission I can try to implement it in your code.

**Comment by boludoz** [comment]
Please, I ask for your authorization and I apologize if I use any of your code without authorization, you can review #11665 and you are free to use my code, you and anyone else, as I ask for your permission to use your code.

**Comment by liamwhite** [comment]
Superseded by #11705

**Pull request closed without merging**